client/core/trade_simnet_test.go

```diff
@@ -0,0 +1,709 @@
+	if err != nil {
+		return fmt.Errorf("client2 place %s order error: %v", sellString(client2.isSeller), err)
+	}
+
+	if finalStatus == order.NewlyMatched {
+		// Lock wallets to prevent Maker from sending swap as soon as the orders are matched.
+		for _, client := range clients {
+			if err = client.lockWallets(); err != nil {
+				return fmt.Errorf("client %d lock wallet error: %v", client.id, err)
+			}
+		}
+	}
+
+	monitorTrades, ctx := errgroup.WithContext(context.Background())
+	monitorTrades.Go(func() error {
+		return monitorTradeForTestOrder(ctx, client1, c1OrderID, finalStatus)
+	})
+	monitorTrades.Go(func() error {
+		return monitorTradeForTestOrder(ctx, client2, c2OrderID, finalStatus)
+	})
+	if err = monitorTrades.Wait(); err != nil {
+		return err
+	}
+
+	tLog.Infof("Trades ended at %s.", finalStatus)
+
+	// Allow some time for balance changes to be properly reported.
+	// There is usually a split-second window where a locked output
+	// has been spent but the spending tx is still in mempool. This
+	// will cause the txout to be included in the wallets locked
+	// balance, causing a higher than actual balance report.
+	time.Sleep(1 * time.Second)
+
+	for _, client := range clients {
+		if err = client.assertBalanceChanges(); err != nil {
+			return fmt.Errorf("client %d balance check error: %v", client.id, err)
+		}
+	}
+
+	return nil
+}
+
+func placeTestOrder(c *tClient, qty, rate uint64, sell bool) (string, error) {
+	dc := c.dc()
+	dcrBtcMkt := dc.market("dcr_btc")
+	if dcrBtcMkt == nil {
+		return "", fmt.Errorf("no dcr_btc market found")
+	}
+	baseAsset := dc.assets[dcrBtcMkt.BaseID]
+	quoteAsset := dc.assets[dcrBtcMkt.QuoteID]
+
+	tradeForm := &TradeForm{
+		Host:    dexHost,
+		Base:    baseAsset.ID,
+		Quote:   quoteAsset.ID,
+		IsLimit: true,
+		Sell:    sell,
+		Qty:     qty,
+		Rate:    rate,
+		TifNow:  false,
+	}
+
+	qtyStr := fmt.Sprintf("%.8f %s", fmtAmt(qty), baseAsset.Symbol)
+	rateStr := fmt.Sprintf("%.8f %s/%s", fmtAmt(rate), quoteAsset.Symbol,
+		baseAsset.Symbol)
+
+	ord, err := c.core.Trade(c.appPass, tradeForm)
+	if err != nil {
+		return "", fmt.Errorf("error placing %s order %v", sellString(sell), err)
+	}
+
+	c.log("placed order %sing %s at %s (%s)", sellString(sell), qtyStr, rateStr, ord.ID[:8])
+	return ord.ID, nil
+}
+
+func monitorTradeForTestOrder(ctx context.Context, client *tClient, orderID string, finalStatus order.MatchStatus) error {
+	errs := newErrorSet("[client %d] ", client.id)
+
+	oid, err := order.IDFromHex(orderID)
+	if err != nil {
+		return errs.add("error parsing order id %s -> %v", orderID, err)
+	}
+
+	oidShort := token(oid.Bytes())
+	client.log("waiting for matches on order %s", oidShort)
+
+	matched := tryUntil(ctx, 15*time.Second, func() bool {
+		select {
+		case n := <-client.notifications:
+			orderNote, isOrderNote := n.(*OrderNote)
+			return isOrderNote && n.Subject() == "Matches made" && orderNote.Order.ID == orderID
```

Comment:
This appears to be the only place you're checking the notification channel. This is fine right now, because the channel is buffered at length 16, but if we were to add more notifications, this might end up causing issues that are difficult to track down. There will be a fair amount of iteration on these tests, I imagine, so it's not super important to do anything about it right now. One thing I'd like to see is to make more judicious use of the notification feed to trigger actions, rather than using `tryUntil`.

Reply:
The notification channel is also checked in the `startClients` function to wait for notice of fee payment confirmation.

> the channel is buffered at length 16, but if we were to add more notifications, this might end up causing issues that are difficult to track down

Are you suggesting not relying on the notification channel to trigger actions?

> One thing I'd like to see is to make more judicious use of the notification feed to trigger actions, rather than using `tryUntil`

That'd be pretty nice. The reason for using `tryUntil` when checking the notification channel is to handle cases where a notification other than the one expected is received from the notification channel. So such unexpected notifications are simply ignored and the wait continues until we get the notification we're looking for or the wait timer expires.

Reply:
It's all good. I'll probably fiddle with this later, though.

Reply:
Btw, I just noticed a series of core error logs: `blocking notification channel`. Do I need to add extra code to keep the channel drained? Is this problematic?

Reply:
That's pretty much what I was getting at. The channel can become blocking and notifications will start being dumped. You aren't relying on notifications for anything else right now, so it's not going to affect your tests passing, but it would be good to keep it drained.